### DISCUSSION TITLE
Updating avatar not working

### DISCUSSION THREAD

**Discussion by itsdereklee** [supabase/supabase#17503] [Questions] [answered]
I'm trying to upload avatar image for the nextjs user management app (provided from supabase docs), it works for the first try and has succesfully stored in storage and has inserted the avatar_url to profiles table.

However the second time i tried to upload, it pops up 'Error updating data', POST 500 error, and has accumulated extra image in the storage without updating the avatar_url in the profiles table (remain the old avatar_url).

<img width="557" alt="image" src="https://github.com/supabase/supabase/assets/84964888/57bd3591-a778-4a8c-b455-daec0f54d887">

I followed exact one-to-one copy from the nextjs user management app doc and has replaced the [delete_storage_object](https://supabase.com/docs/guides/getting-started/tutorials/with-angular#storage-management) with my project_url and service_role_key.

I've checked github discussion etc, seems like other is having the same issue as well.

**Comment by GaryAustin1** [comment]
You should look in the Postgres log of the dashboard for the database error specifics.  Likely that error is coming from the database.

**Reply by itsdereklee** [reply]
@GaryAustin1 Hi, thanks for the reply. I have received this error from the postgres log:

![image](https://github.com/supabase/supabase/assets/84964888/8df583ad-844f-456b-b7d6-273ab4baf33f)

As I am learning supabase, what is exact issue for it?

**Reply by GaryAustin1** [reply]
Your code for some reason is not providing the url to the http extension.
For example this url is for some reason null.  I can't see how with the example code though. If that is even what your code looks like or if there is another http extension call somewhere.
![image](https://github.com/supabase/supabase/assets/54564956/6b019d6b-0631-4db2-99d7-3008dcb035c7)

**Reply by itsdereklee** [reply]
Could it be the supabase doc error? because I had enabled http extension and done everything as mentioned in the documents. However, as I continue to search for answer, other users seem to have similar issue as well (http request return null).

Or perhaps is there any fix for this issue?

**Reply by GaryAustin1** [reply]
I don't think the code in the doc is in error or at least the function doing the delete with the http call.

I ran that code in the SQL editor and it deleted the file...
```
do $$
declare
  status int;
  content text;
  project_url text := 'https://xxxxxxxxxxxxxxxxxxxx.supabase.co';
  service_role_key text := 'myservicerole'; --  full access needed
  url text := project_url||'/storage/v1/object/'||'test'||'/'|| 'public/test49.jpg';
begin
  select
      into status, content
           result.status::int, result.content::text
      FROM extensions.http((
    'DELETE',
    url,
    ARRAY[extensions.http_header('authorization','Bearer '||service_role_key)],
    NULL,
    NULL)::extensions.http_request) as result;
end;
$$;
```

I'm not really sure what could cause it outside of that code or if permissions where messed up.  I see 2 other people reporting uri null also.

**Reply by itsdereklee** [reply]
Did it manage to change the avatar_url in the profiles table? After uploading a picture the second time and so fourth...because mine remains the same avatar_url.

If that's the case then there might be some issue with the permission settings.

**Reply by GaryAustin1** [reply]
I just ran that bit of code which only does the delete file API call.
So far the only way I get the URI is null error is if I set the url in the code to NULL.  Anything else for the parameters the string concat function return uri bad format errors.

I would expect a permission or not found error on something like if say extensions.http_request() did not work.   So at the moment I'm not sure what could cause it, but I don't see how the url variable can get to null with the code that is there.  I don't use the guides, but is there possibly another http extension call elsewhere?  I did not see one in the guide with a text search.

Edit:  I also made the call http(null) to simulate http_extension somehow returning null and that is a different error.

**Reply by GaryAustin1** [reply]
You could try running the sample like I did in the SQL editor.  The function in the guide was declared security definer type so should not have any permission issues different than the SQL editor.

**Reply by itsdereklee** [reply]
Hi Gary, you're right. I've tried your code and it successfully deleted my pictures. However, at this point I am sort of lost. Although it successfully deleted the pictures, I am not sure what is the next move that I should do to fix the update of avatar_url, after user had updated/upload their new profile picture (which would post new avatar_url and replace the old avatar_url) to display it out at the homepage.

<img width="883" alt="image" src="https://github.com/supabase/supabase/assets/84964888/8dab3014-de4e-4ac3-b64b-3c14d13b3af2">

**Reply by itsdereklee** [reply]
The code above provided by supabase doc does not work as mentioned (I believed this is the part where other users is having issue as well). There is no warning appears in the log either, except POST 500 and http request null error, from the above reply.

**Reply by GaryAustin1** [reply]
The code you just showed SHOULD not be involved in the error as it calls the function we have been discussing.  If anything fails in any part of the trigger functions they all fail.  The error is pretty specific as it says http_request uri is null.  Unless you have another http extension call somewhere I assume it is the piece of code we were looking at.

**Reply by itsdereklee** [reply]
At the moment I'm not sure what could cause it, but I don't see how the url variable can get to null with the code that is there. There is no other http extension call either from the docs.

**Reply by itsdereklee** [reply]
I figured out a solution, and that is this code delete_old_avatar trigger is messing up with [delete_storage_object](https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs#storage-management) function.

What i did was to simply just remove this code and the avatar_url updated itself, and the profile picture has been updated successfully (second try, third try and so fourth). However, it accumulated extra image in the storage. Wonder if there is any fix around for this? @GaryAustin1

**Comment by itsdereklee** [comment, marked as answer]
Hello everyone! The support team has confirmed that there are mistakes in the documentation's code. In order to resolve the issue please amend the trigger `delete_old_avatar` with the following snippet.

```
-- Create or replace the function 'delete_old_avatar'
CREATE OR REPLACE FUNCTION delete_old_avatar()
RETURNS TRIGGER
LANGUAGE 'plpgsql'
SECURITY DEFINER
AS $$
DECLARE
    -- Declare variables
    status INT;
    content TEXT;
    avatar_name TEXT;

BEGIN
    -- Check if the old avatar URL is not empty and if the operation is DELETE or avatar URL changed
    IF COALESCE(old.avatar_url, '') <> ''
       AND (tg_op = 'DELETE' OR (old.avatar_url <> new.avatar_url)) THEN

        -- Extract the avatar name from the old avatar URL
        avatar_name := old.avatar_url;

        -- Call the public.delete_avatar function and store the result
        SELECT INTO status, content
               result.status, result.content
        FROM public.delete_avatar(avatar_name) AS result;

        -- Raise a warning if the status is not 200 (OK)
        IF status <> 200 THEN
            RAISE WARNING 'Could not delete avatar: % %', status, content;
        END IF;
    END IF;

    -- Return the old record for DELETE operations
    IF tg_op = 'DELETE' THEN
        RETURN old;
    END IF;

    -- Return the new record for other operations
    RETURN new;

END;
$$;
```

To give more details, the issue was located during the `avatar_name := ` variable initialisation. The documentation sets the variable as following:

`avatar_name := substring(old.avatar_url from '/([^\/]+)\?.*$');`

Which tries to strip the bucket name out of the file name. However, the frontend is not storing the bucket name within the avatar_url field but simply the file name. For this reason, if we remove the substring function and simply assign it as following:

`avatar_name := old.avatar_url;`

All will work as expected.

**Reply by EasternIndustries-IT** [reply]
This unfortunately did not solve my issue. I am having the exact same issue (can upload once but not a 2nd or 3rd time). Following the [tutorial.](https://supabase.com/docs/guides/getting-started/tutorials/with-vue-3?database-method=sql) But I'm getting a connection refused. The image is uploaded but it is refusing to change the database name for avatar_url.

![image](https://github.com/supabase/supabase/assets/57259898/80917f4a-886a-482b-a184-2934443d81e1)

![image](https://github.com/supabase/supabase/assets/57259898/023d15dd-2965-4185-9f32-281a1b02085b)

This is what shows in the database under PostgREST Log
`172.18.0.13 - authenticated [20/Oct/2023:20:36:14 +0000] "POST /profiles HTTP/1.1" 500 - "http://localhost:7000/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36"`

I don't believe this is an authentication issue because it worked the first time. Not sure why it would connect once then fail after. Any ideas?
